### PR TITLE
Improve rockcraft file for dhcp_relay

### PR DESCRIPTION
**Pull request by jy5275**
This PR improves rockcraft file for dhcp_relay:
- Merge the `start.sh` into `init.sh`
- Add a python script `gen_services.py` to generate dynamic service config (previously supervisord did this by jinjia2)
- Remove supervisord packages and configs

### CODE REVIEW

**Comment by jy5275** [comment]
Hi @benhoyt, this PR is ready to merge. 
We have to merge this PR with [this PR](https://github.com/canonical/sonic-buildimage/pull/30) together to solve the CI failing issue.

**Comment by benhoyt** [comment]
@jy5275 Understood. No rush on this at all, let's wait till after Chinese New Year and we're back from the mini-sprint.

**Comment by jy5275** [comment]
Hi @benhoyt this is ready to merge

File: dockers/docker-dhcp-relay/rock_init.sh [hunk 1]
```diff
@@ -17,3 +23,35 @@ sonic-cfggen $CFGGEN_PARAMS
...
+    # on is down, the relay agent will not start. If an interface the relay
+    # should listen on is up but does not have an IP address assigned when the
+    # relay agent starts, it will not listen or send on that interface for the
+    # lifetime of the process.
+    /usr/bin/wait_for_intf.sh
+fi
+
+pebble start dhcprelayd
+
+python3 /usr/bin/gen_services.py
```
Comment (benhoyt) [inline dockers/docker-dhcp-relay/rock_init.sh:57]: I would much rather the gen_services.py script just generated the config files, and then we did the `pebble replan` separately here.

File: dockers/docker-dhcp-relay/gen_services.py [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+    cmd_reload = ["pebble", "replan"]
+    subprocess.run(cmd_reload, check=True)
+
+    # Start all services in the new layer
+    first_pass, second_pass = get_start_order(pebble_conf.get("services", {}).keys())
+    for svc in first_pass:
+        cmd_start = ["pebble", "start", svc]
+        subprocess.run(cmd_start, check=True)
+        
+    for svc in second_pass:
```
Comment (benhoyt) [inline dockers/docker-dhcp-relay/gen_services.py:39]: Is splitting into first-pass and second-pass needed? Given that `pebble start` only waits 1s and then returns success, what is gained by doing these in two passes? Also, Pebble can start multiple services in a single `pebble start` invocation -- should we use that?

File: dockers/docker-dhcp-relay/gen_services.py [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+
+import json
+import subprocess
+import sys
+import ipaddress
+import yaml
+from typing import Dict, Any, Iterable, Tuple
+
+LAYER_NAME = "dynamic_services"
+TMP_CONFIG_FILE_PATH = "/tmp/pebble_dynamic_services.yaml"
```
Comment (benhoyt) [inline dockers/docker-dhcp-relay/gen_services.py]: It's better to use Python's `tempfile` module instead of hard-coding `/tmp`. Probably `with tempfile.TemporaryFile(...)`

File: dockers/docker-dhcp-relay/gen_services.py [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+        subprocess.run(cmd_start, check=True)
+
+
+def fetch_cfg_var(var_name: str) -> Any:
+    # Run `sonic-cfggen -d --var-json VAR` and return parsed JSON, or None on error.
+    cmd = ["sonic-cfggen", "-d", "--var-json", var_name]
+    try:
+        out = subprocess.check_output(cmd, text=True)
+        return json.loads(out)
+    except Exception:
```
Comment (benhoyt) [inline dockers/docker-dhcp-relay/gen_services.py:50]: It's almost never good to catch all exceptions; rather use a specific exception type, otherwise even typos will become AttributeError or NameError, for example. In this case, use `subprocess.CalledProcessError` if the process will fail, or a `try/except ValueError` around just the `json.loads` if that is what will fail.

File: dockers/docker-dhcp-relay/gen_services.py [hunk 5]
```diff
@@ -0,0 +1,316 @@
...
+
+def is_ipv4(addr: str) -> bool:
+    if not addr:
+        return False
+    # strip any surrounding whitespace and CIDR
+    a = str(addr).strip().split("/")[0]
+    try:
+        ipaddress.IPv4Address(a)
+        return True
+    except Exception:
```
Comment (benhoyt) [inline dockers/docker-dhcp-relay/gen_services.py:62]: Similar here. Use `ValueError` for this case (see the `ipaddress` docs).

File: dockers/docker-dhcp-relay/gen_services.py [hunk 6]
```diff
@@ -0,0 +1,316 @@
...
+        return json.loads(out)
+    except Exception:
+        return None
+
+
+def is_ipv4(addr: str) -> bool:
+    if not addr:
+        return False
+    # strip any surrounding whitespace and CIDR
+    a = str(addr).strip().split("/")[0]
```
Comment (benhoyt) [inline dockers/docker-dhcp-relay/gen_services.py:58]: Why `str()` when it's already a string? Why strip? I think we should use `ipaddress.IPv4Network()` here to parse instead of doing ad-hoc parsing ourselves, and then check `network.prefixlen` or `network.network_address` or similar. What about something like:

**Review by benhoyt — COMMENTED** [review]
I'd like to see if we can figure out a better way to do this. I started reviewing the Python code, and there are a lot of small quality issues with it (and I think it's over-complicated for what we need) so I didn't finish reviewing it.

Here we're adding 300 lines of custom Python with lots of if/elses and edge cases, and only a single small test. In addition, we're deviating from upstream a lot. If we're going to harden this product, all this code needs to be production-quality and reviewed thoroughly. Which we can do if absolutely necessary, but I want to try to find alternatives first. Let's look at this and the source files more closely in our 1:1 tomorrow.
